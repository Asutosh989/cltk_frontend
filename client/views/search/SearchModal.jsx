import baseTheme from 'material-ui/styles/baseThemes/lightBaseTheme';
import IconButton from 'material-ui/IconButton';
import getMuiTheme from 'material-ui/styles/getMuiTheme';


SearchModal = React.createClass({

	propTypes: {
		closeSearchModal: React.PropTypes.func,
		visible: React.PropTypes.bool,
	},

	mixins: [ReactMeteorData],

	childContextTypes: {
		muiTheme: React.PropTypes.object.isRequired,
	},

	getInitialState() {
		return {
			layout: 'grid',
			filters: [],
			skip: 0,
			limit: 12,
		};
	},

	getChildContext() {
		return { muiTheme: getMuiTheme(baseTheme) };
	},


	getMeteorData() {
		const query = {};
		let works = [];
		let stillMoreWorks = true;

		// Parse the filters to the query
		this.state.filters.forEach((filter) => {
			const date = moment(`${filter.values[0]}-01-01`, 'YYYY MM DD');
			switch (filter.key) {
			case 'textsearch':
				query.$text = { $search: filter.values[0] };
				break;

			case '':
				query.$or = [{$where: "this.miradorLink && this.miradorLink.length > 0"}, {hasImageViewer: true}];
				break;

			case 'scribes':
				query.scribe = { $in: filter.values };
				break;

			case 'illuminators':
				query.illuminator = { $in: filter.values };
				break;

			case 'institutions':
				query.institution = { $in: filter.values };
				break;

			case 'places':
				query.place = { $in: filter.values };
				break;

			case 'dateFrom':
				query.dateBegun = { $gte: new Date(date.toISOString()) };
				break;

			case 'dateTo':
				query.dateEnded = { $lte: new Date(date.toISOString()) };
				break;
			default:
				// do nothing
			}
		});

		console.log('Works query:', query);
		const handle = Meteor.subscribe('works', query, this.props.skip, this.props.limit);
		if (handle.ready()) {
			works = Works.find({}, {}).fetch();
			/*
			objects.forEach((object, i) => {
				const imageSubscription = Meteor.subscribe('objectImages', object.slug);
				if (imageSubscription.ready()) {
					objects[i].images = Images.find({}).fetch();
					objects[i].thumbnails = Thumbnails.find({}).fetch();
				}
			});
			*/

			if (works.length < this.props.limit) {
				stillMoreWorks = false;
			}
		}

		return {
			works,
			stillMoreWorks,
		};
	},

	loadMoreWorks() {
		console.log('SearchModal.loadMoreWorks', this.state.skip + this.state.limit);
		this.setState({
			skip: this.state.skip + this.state.limit,
		});
	},

	toggleSearchTerm(key, value) {
		const filters = this.state.filters;
		let keyIsInFilter = false;
		let valueIsInFilter = false;
		let filterValueToRemove;
		let filterToRemove;

		filters.forEach((filter, i) => {
			if (filter.key === key) {
				keyIsInFilter = true;

				if (filter.values.indexOf(value) >= 0) {
					valueIsInFilter = true;
					filterValueToRemove = filter.values.indexOf(value);
				}

				if (valueIsInFilter) {
					filter.values.splice(filterValueToRemove, 1);
					if (filter.values.length === 0) {
						filterToRemove = i;
					}
				} else {
					filter.values.push(value);
				}
			}
		});


		if (typeof filterToRemove !== 'undefined') {
			filters.splice(filterToRemove, 1);
		}

		if (!keyIsInFilter) {
			filters.push({
				key,
				values: [value],
			});
		}

		this.setState({
			filters,
			skip: 0,
		});
	},

	handleChangeTextsearch(textsearch) {
		const filters = this.state.filters;

		if (textsearch && textsearch.length) {
			let textsearchInFilters = false;
			console.log(filters);

			filters.forEach((filter, i) => {
				if (filter.key === 'textsearch') {
					filters[i].values = [textsearch];
					textsearchInFilters = true;
				}
			});
			console.log(filters);

			if (!textsearchInFilters) {
				filters.push({
					key: 'textsearch',
					values: [textsearch],
				});
			}
		} else {
			let filterToRemove;

			filters.forEach((filter, i) => {
				if (filter.key === 'textsearch') {
					filterToRemove = i;
				}
			});

			if (typeof filterToRemove !== 'undefined') {
				filters.splice(filterToRemove, 1);
			}
		}

		this.setState({
			filters,
		});
	},

	handleChangeDate(e) {
		const filters = this.state.filters;

		let dateFromInFilters = false;

		filters.forEach((filter, i) => {
			if (filter.key === 'dateFrom') {
				filters[i].values = [e.from];
				dateFromInFilters = true;
			}
		});

		if (!dateFromInFilters) {
			filters.push({
				key: 'dateFrom',
				values: [e.from],
			});
		} else {
			let filterToRemove;

			filters.forEach((filter, i) => {
				if (filter.key === 'dateFrom') {
					filterToRemove = i;
				}
			});

			if (typeof filterToRemove !== 'undefined') {
				filters.splice(filterToRemove, 1);
			}
		}

		let dateToInFilters = false;

		filters.forEach((filter, i) => {
			if (filter.key === 'dateTo') {
				filters[i].values = [e.to];
				dateToInFilters = true;
			}
		});

		if (!dateToInFilters) {
			filters.push({
				key: 'dateTo',
				values: [e.to],
			});
		} else {
			let filterToRemove;

			filters.forEach((filter, i) => {
				if (filter.key === 'dateTo') {
					filterToRemove = i;
				}
			});

			if (typeof filterToRemove !== 'undefined') {
				filters.splice(filterToRemove, 1);
			}
		}

		this.setState({
			filters,
		});
	},

	toggleLayout(layout) {
		this.setState({
			layout,
		});
	},


	render() {
		return (
			<div className={'cltk-modal search-modal ' + (this.props.visible ? 'search-modal--visible' : '')} >
				<div className="close-search">
					<IconButton
						iconClassName={'close-search-icon mdi mdi-close'}
						onClick={this.props.closeSearchModal}
						onTouchTap={this.props.closeSearchModal}
						/>
				</div>

				{this.props.visible ?
					<div>
						<SearchTools
							filters={this.state.filters}
							toggleSearchTerm={this.toggleSearchTerm}
							handleChangeDate={this.handleChangeDate}
							handleChangeTextsearch={this.handleChangeTextsearch}
						/>
						<SearchFilters
							filters={this.state.filters}
							toggleSearchTerm={this.toggleSearchTerm}
						/>

						<section className="search-results">
							<SearchResultsList
								works={this.data.works}
								/>

						</section>

					</div>
					: ""
				}


			</div>
		);
	},
});
