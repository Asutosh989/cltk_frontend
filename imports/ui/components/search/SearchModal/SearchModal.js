import React from 'react';
import PropTypes from 'prop-types';
import { graphql, gql } from 'react-apollo';
import autoBind from 'react-autobind';
import baseTheme from 'material-ui/styles/baseThemes/lightBaseTheme';
import getMuiTheme from 'material-ui/styles/getMuiTheme';
import IconButton from 'material-ui/IconButton';

import Authors from '/imports/api/collections/authors';
import Works from '/imports/api/collections/works';

import SearchTools from '/imports/ui/components/search/SearchTools';
import SearchFilters from '/imports/ui/components/search/SearchFilters';
import SearchResultsList from '/imports/ui/components/search/SearchResultsList';

class SearchModal extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			layout: 'grid',
		};

		autoBind(this);
	}

	getChildContext() {
		return { muiTheme: getMuiTheme(baseTheme) };
	}

	loadMoreWorks() {
		const { offset, limit } = this.props;
		this.props.changeSearchParams({
			offset: offset + limit,
		});
	}

	toggleSearchTerm(key, value) {
		const { filters } = this.props;
		let keyIsInFilter = false;
		let valueIsInFilter = false;
		let filterValueToRemove;
		let filterToRemove;

		filters.forEach((filter, i) => {
			if (filter.key === key) {
				keyIsInFilter = true;

				if (
						key === 'authors'
						&& filter.values.some((existingValue) =>
							existingValue._id === value._id
						)) {
					valueIsInFilter = true;
					filterValueToRemove = filter.values.indexOf(value);
				} else if (filter.values.indexOf(value) >= 0) {
					valueIsInFilter = true;
					filterValueToRemove = filter.values.indexOf(value);
				}

				if (valueIsInFilter) {
					filter.values.splice(filterValueToRemove, 1);
					if (filter.values.length === 0) {
						filterToRemove = i;
					}
				} else {
					filter.values.push(value);
				}
			}
		});


		if (typeof filterToRemove !== 'undefined') {
			filters.splice(filterToRemove, 1);
		}

		if (!keyIsInFilter) {
			filters.push({
				key,
				values: [value],
			});
		}

		this.props.changeSearchParams({
			filters,
			offset: 0,
		});
	}

	handleChangeTextsearch(textsearch) {
		const { filters } = this.props;

		if (textsearch && textsearch.length) {
			let textsearchInFilters = false;

			filters.forEach((filter, i) => {
				if (filter.key === 'textsearch') {
					filters[i].values = [textsearch];
					textsearchInFilters = true;
				}
			});

			if (!textsearchInFilters) {
				filters.push({
					key: 'textsearch',
					values: [textsearch],
				});
			}
		} else {
			let filterToRemove;

			filters.forEach((filter, i) => {
				if (filter.key === 'textsearch') {
					filterToRemove = i;
				}
			});

			if (typeof filterToRemove !== 'undefined') {
				filters.splice(filterToRemove, 1);
			}
		}

		this.props.changeSearchParams({
			filters,
			offset: 0,
		});
	}

	handleChangeDate(e) {
		const { filters } = this.props;

		let dateFromInFilters = false;

		filters.forEach((filter, i) => {
			if (filter.key === 'dateFrom') {
				filters[i].values = [e.from];
				dateFromInFilters = true;
			}
		});

		if (!dateFromInFilters) {
			filters.push({
				key: 'dateFrom',
				values: [e.from],
			});
		} else {
			let filterToRemove;

			filters.forEach((filter, i) => {
				if (filter.key === 'dateFrom') {
					filterToRemove = i;
				}
			});

			if (typeof filterToRemove !== 'undefined') {
				filters.splice(filterToRemove, 1);
			}
		}

		let dateToInFilters = false;

		filters.forEach((filter, i) => {
			if (filter.key === 'dateTo') {
				filters[i].values = [e.to];
				dateToInFilters = true;
			}
		});

		if (!dateToInFilters) {
			filters.push({
				key: 'dateTo',
				values: [e.to],
			});
		} else {
			let filterToRemove;

			filters.forEach((filter, i) => {
				if (filter.key === 'dateTo') {
					filterToRemove = i;
				}
			});

			if (typeof filterToRemove !== 'undefined') {
				filters.splice(filterToRemove, 1);
			}
		}

		this.props.changeSearchParams({
			filters,
			offset: 0,
		});
	}

	toggleLayout(layout) {
		this.setState({
			layout,
		});
	}

	closeSearchModal() {
		this.props.changeSearchParams({ filters: [] });
		this.props.closeSearchModal();
	}

	render() {
		let hasMoreWorks = true;
		const { works, worksCount, limit } = this.props;

		if (works) {
			if (worksCount && worksCount <= works.length) {
				hasMoreWorks = false;
			} else if (works.length < limit) {
				hasMoreWorks = false;
			}
		}

		return (
			<div
				className={`cltk-modal search-modal
				${(this.props.visible ? 'search-modal--visible' : '')}`}
			>
				<div className="close-search">
					<IconButton
						iconClassName={'close-search-icon mdi mdi-close'}
						onClick={this.closeSearchModal}
						onTouchTap={this.closeSearchModal}
					/>
				</div>

				{this.props.visible ?
					<div>
						<SearchTools
							filters={this.props.filters}
							toggleSearchTerm={this.toggleSearchTerm}
							handleChangeDate={this.handleChangeDate}
							handleChangeTextsearch={this.handleChangeTextsearch}
						/>
						<SearchFilters
							filters={this.props.filters}
							toggleSearchTerm={this.toggleSearchTerm}
						/>
						<section className="search-results">
							<SearchResultsList
								works={works}
								hasMoreWorks={hasMoreWorks}
								loadMore={this.loadMoreWorks}
							/>
						</section>
					</div>
					: ''
				}
			</div>
		);
	}
};

SearchModal.childContextTypes = {
	muiTheme: PropTypes.object.isRequired,
};

SearchModal.propTypes = {
	changeSearchParams: PropTypes.func.isRequired,
	closeSearchModal: PropTypes.func,
	filters: PropTypes.array,
	limit: PropTypes.number,
	offset: PropTypes.number,
	visible: PropTypes.bool,
	work: PropTypes.object,
};

const withData = graphql(gql`{
  works {
		id
		english_title
		original_title
		slug
		author
		corpus
	}
	corpora {
		id
		slug
		title
	}
	languages {
		id
		slug
		title
	}
	worksCount
}`, {
  options: ({ filters, offset, limit } ) => {
		const query = {};

		// Parse the filters to the query
		filters.forEach((filter) => {
			const date = moment(`${filter.values[0]}-01-01`, 'YYYY MM DD');
			switch (filter.key) {
			case 'textsearch': {
				query.textsearch = filter.values[0];
				break;
			}

			case 'languages': {
				query.languages = filter.values;
				break;
			}

			case 'corpora': {
				query.corpora = filter.values;
				break;
			}

			case 'authors': {
				const values = [];

				filter.values.forEach((value) => {
					values.push(value.id);
				});

				query.authors = values;
				break;
			}

			case 'dateStart': {
				query.dateBegun = new Date(date.toISOString());
				break;
			}

			case 'dateEnd': {
				query.dateEnded = new Date(date.toISOString());
				break;
			}

			default: {
				// do nothing
				break;
			}
			}
		});

		return {
	    variables: {
				...query,
				limit,
				offset,
			},
		};
  },
  props: ({ data: { works, worksCount, uniqueCorpora, uniqueLanguages } }) => ({
		works, worksCount, uniqueCorpora, uniqueLanguages,
  }),
});

export default withData(SearchModal);
