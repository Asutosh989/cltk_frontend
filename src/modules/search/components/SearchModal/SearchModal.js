import React from 'react';
import PropTypes from 'prop-types';
import { graphql, gql } from 'react-apollo';
import moment from 'moment';
import autoBind from 'react-autobind';
import baseTheme from 'material-ui/styles/baseThemes/lightBaseTheme';
import getMuiTheme from 'material-ui/styles/getMuiTheme';
import IconButton from 'material-ui/IconButton';

import Authors from '/imports/api/collections/authors';
import Works from '/imports/api/collections/works';

import SearchTools from '/imports/ui/components/search/SearchTools';
import SearchFilters from '/imports/ui/components/search/SearchFilters';
import SearchResultsList from '/imports/ui/components/search/SearchResultsList';

class SearchModal extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			layout: 'grid',
		};

		autoBind(this);
	}

	getChildContext() {
		return { muiTheme: getMuiTheme(baseTheme) };
	}

	handleChangeTextsearch(textsearch) {
		this.props.handleChangeTextsearch({
			textsearch,
			offset: 0,
			limit: 21,
		});
	}

	toggleLayout(layout) {
		this.setState({
			layout,
		});
	}

	closeSearchModal() {
		this.props.closeSearchModal();
	}

	render() {
		let hasMoreWorks = true;
		const { works, worksCount, limit } = this.props;

		if (works) {
			if (worksCount && worksCount <= works.length) {
				hasMoreWorks = false;
			} else if (works.length < limit) {
				hasMoreWorks = false;
			}
		}

		return (
			<div
				className={`cltk-modal search-modal
				${(this.props.visible ? 'search-modal--visible' : '')}`}
			>
				<div className="close-search">
					<IconButton
						iconClassName={'close-search-icon mdi mdi-close'}
						onClick={this.closeSearchModal}
						onTouchTap={this.closeSearchModal}
					/>
				</div>

				{this.props.visible ?
					<div>
						<SearchTools
							filters={this.props.filters}
							toggleSearchTerm={this.toggleSearchTerm}
							handleChangeDate={this.handleChangeDate}
							handleChangeTextsearch={this.handleChangeTextsearch}
						/>
						<section className="search-results">
							<SearchResultsList
								works={works}
								hasMoreWorks={hasMoreWorks}
								loadMore={this.props.loadMore}
							/>
						</section>
					</div>
					: ''
				}
			</div>
		);
	}
};

SearchModal.childContextTypes = {
	muiTheme: PropTypes.object.isRequired,
};

SearchModal.propTypes = {
	handleChangeTextsearch: PropTypes.func.isRequired,
	closeSearchModal: PropTypes.func,
	textsearch: PropTypes.string,
	limit: PropTypes.number,
	offset: PropTypes.number,
	visible: PropTypes.bool,
	work: PropTypes.object,
};

const withData = graphql(gql`
  query SearchWorksWithTitle($textsearch: String!){
		search_works_by_title(title: $textsearch) {
			id
			english_title
			original_title
			slug
			language {
				title
			}
			author {
				id
				name
			}
			corpus {
				id
				title
				slug
			}
		}
		corpora {
			id
			slug
			title
		}
		languages {
			id
			slug
			title
		}
		works_count
	}`, {
		options: ({ textsearch, limit, offset }) => {
			return {
	    variables: {
		textsearch,
		limit,
		offset,
	},
			};
		},
		props: ({ data: { search_works_by_title, works_count, uniqueCorpora, uniqueLanguages } }) => ({
			works: search_works_by_title,
			worksCount: works_count,
			uniqueCorpora,
			uniqueLanguages,
		}),
	});

export default withData(SearchModal);
