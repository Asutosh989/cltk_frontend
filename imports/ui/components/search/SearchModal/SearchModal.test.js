
import React from 'react';
import renderer from 'react-test-renderer';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';

// component:
import SearchModal from './SearchModal';

describe('SearchModal', () => {
	it('renders correctly', () => {

		const tree = renderer
			.create(
				<MuiThemeProvider>
					<SearchModal />
				</MuiThemeProvider>
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
