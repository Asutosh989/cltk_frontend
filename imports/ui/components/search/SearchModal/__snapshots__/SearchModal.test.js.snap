// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchModal renders correctly 1`] = `
<div
  className="cltk-modal search-modal
				"
>
  <div
    className="close-search"
  >
    <button
      disabled={false}
      href={undefined}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseOut={[Function]}
      onTouchStart={[Function]}
      onTouchTap={[Function]}
      style={
        Object {
          "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
          "background": "none",
          "border": 10,
          "boxSizing": "border-box",
          "cursor": "pointer",
          "display": "inline-block",
          "fontFamily": "Roboto, sans-serif",
          "fontSize": 0,
          "fontWeight": "inherit",
          "height": 48,
          "margin": 0,
          "muiPrepared": true,
          "outline": "none",
          "overflow": "visible",
          "padding": 12,
          "position": "relative",
          "textDecoration": "none",
          "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
          "verticalAlign": null,
          "width": 48,
          "zIndex": 1,
        }
      }
      tabIndex={0}
      type="button"
    >
      <div
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
      >
        <span
          className="close-search-icon mdi mdi-close"
          color="rgba(0, 0, 0, 0.87)"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "color": "rgba(0, 0, 0, 0.87)",
              "display": "inline-block",
              "fontSize": 24,
              "muiPrepared": true,
              "position": "relative",
              "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              "userSelect": "none",
            }
          }
        />
      </div>
    </button>
  </div>
  
</div>
`;
